File: ios/SwishStrikeCore/Sources/SwishStrikeCore/CountingEngine.swift
import Foundation

// ----------------------------------------------------------------------------
// Swish Strike — CountingEngine (Swift)
//
// The Swift mirror of web-prototype/js/countingEngine.js. Same input sample
// sequence MUST produce the same count. This is the brain of the iOS app and it
// is intentionally pure — no camera, no Vision, no UI — so it unit-tests with
// `swift run SwishStrikeCoreCheck`. Parity with the JS engine is guarded by matched
// suites AND a shared golden trace (golden.fixtures.json) replayed through both.
//
// Robustness guards (see update()): out-of-order/duplicate frames are dropped,
// non-finite coordinates are treated as misses, a long detector gap resets the
// track so no phantom crossing is synthesized across an occlusion, and a
// degenerate (zero-area) zone never fires.
// ----------------------------------------------------------------------------

public enum RuleType: String, Sendable {
    case zoneCrossDown
    case zoneStreak
    case bounceReversal
}

/// A target zone in normalized coordinates (0...1, y pointing down).
public struct Zone: Sendable, Equatable {
    public var left, top, right, bottom: Double
    public init(left: Double, top: Double, right: Double, bottom: Double) {
        self.left = left; self.top = top; self.right = right; self.bottom = bottom
    }
}

public enum ReversalDirection: String, Sendable {
    case bottom // troughs (juggle touch, floor bounce)
    case top    // apexes
}

public struct CountRule: Sendable {
    public var type: RuleType
    public var zone: Zone?
    public var xTolerance: Double
    public var armWindow: Double
    public var cooldown: Double
    public var direction: ReversalDirection
    public var minAmplitude: Double
    public var missMargin: Double // zoneStreak: how far below the zone counts as a miss
    public var smoothingAlpha: Double
    public var minConfidence: Double
    public var maxGap: Double

    public static func zoneCrossDown(_ zone: Zone,
                                     xTolerance: Double = 0.05,
                                     armWindow: Double = 1.5,
                                     cooldown: Double = 1.0,
                                     smoothingAlpha: Double = 0.5,
                                     minConfidence: Double = 0.30,
                                     maxGap: Double = 1.5) -> CountRule {
        CountRule(type: .zoneCrossDown, zone: zone, xTolerance: xTolerance,
                  armWindow: armWindow, cooldown: cooldown, direction: .bottom,
                  minAmplitude: 0, missMargin: 0.18, smoothingAlpha: smoothingAlpha,
                  minConfidence: minConfidence, maxGap: maxGap)
    }

    /// zoneStreak rule. Detects makes exactly like zoneCrossDown, but the engine count
    /// tracks the CONSECUTIVE streak: a make increments it, and a miss (the ball, after
    /// being aimed at the target, falls more than `missMargin` past the zone without
    /// scoring) resets it to 0. Used by Free-Throw Streak.
    public static func zoneStreak(_ zone: Zone,
                                  xTolerance: Double = 0.05,
                                  armWindow: Double = 1.5,
                                  cooldown: Double = 1.0,
                                  missMargin: Double = 0.18,
                                  smoothingAlpha: Double = 0.5,
                                  minConfidence: Double = 0.30,
                                  maxGap: Double = 1.5) -> CountRule {
        CountRule(type: .zoneStreak, zone: zone, xTolerance: xTolerance,
                  armWindow: armWindow, cooldown: cooldown, direction: .bottom,
                  minAmplitude: 0, missMargin: missMargin, smoothingAlpha: smoothingAlpha,
                  minConfidence: minConfidence, maxGap: maxGap)
    }

    public static func bounceReversal(direction: ReversalDirection = .bottom,
                                      minAmplitude: Double = 0.12,
                                      cooldown: Double = 0.25,
                                      smoothingAlpha: Double = 0.5,
                                      minConfidence: Double = 0.30,
                                      maxGap: Double = 1.5) -> CountRule {
        CountRule(type: .bounceReversal, zone: nil, xTolerance: 0,
                  armWindow: 0, cooldown: cooldown, direction: direction,
                  minAmplitude: minAmplitude, missMargin: 0.18, smoothingAlpha: smoothingAlpha,
                  minConfidence: minConfidence, maxGap: maxGap)
    }
}

/// One detection sample. A missing detection = nil x/y, or confidence below floor.
public struct Sample: Sendable {
    public var t: Double
    public var x: Double?
    public var y: Double?
    public var confidence: Double?
    public init(t: Double, x: Double? = nil, y: Double? = nil, confidence: Double? = nil) {
        self.t = t; self.x = x; self.y = y; self.confidence = confidence
    }
}

public struct ScoreEvent: Sendable, Equatable {
    public var t: Double
    public var count: Int
    public var quality: String?     // "swish" | "rim" — shot-quality metadata (zoneCrossDown)
    public var centerError: Double? // 0 = dead center, 1 = at the zone edge
    public var type: String?        // "miss" on a streak-reset event; nil for a make
    public init(t: Double, count: Int, quality: String? = nil, centerError: Double? = nil, type: String? = nil) {
        self.t = t; self.count = count; self.quality = quality; self.centerError = centerError; self.type = type
    }
}

public final class CountingEngine {
    public let rule: CountRule
    public private(set) var count: Int = 0
    public private(set) var events: [ScoreEvent] = []
    public private(set) var lastEvent: ScoreEvent?  // most recent score event (incl. shot quality)
    public private(set) var justMissed = false      // true only on the update where a streak miss resolved

    private var sx: Double?
    private var sy: Double?
    private var py: Double?
    private var lastT: Double = -.infinity
    private var lastValidT: Double = -.infinity
    private var lastCountT: Double = -.infinity
    private var lastQuality: (quality: String, centerError: Double)? // computed at fire, consumed on append
    // zoneCrossDown
    private var armed = false
    private var armedT: Double = -.infinity
    // zoneStreak: a shot is in flight (armed once, awaiting make-or-miss resolution)
    private var attemptActive = false
    // shot-quality (swish vs rim)
    private var descentReversals = 0
    private var lastDySign = 0
    // bounceReversal
    private var extremeVal: Double?
    private var sawApproach = false

    public init(rule: CountRule) { self.rule = rule }

    public func reset() {
        count = 0
        events.removeAll()
        lastEvent = nil
        justMissed = false
        lastT = -.infinity
        lastValidT = -.infinity
        lastCountT = -.infinity
        lastQuality = nil
        resetTrack()
    }

    /// Clears only the per-frame tracking state (not the count/history).
    private func resetTrack() {
        sx = nil; sy = nil; py = nil
        armed = false; armedT = -.infinity
        attemptActive = false
        descentReversals = 0; lastDySign = 0
        extremeVal = nil; sawApproach = false
    }

    /// Smoothed position, or nil if nothing has been seen yet.
    public var position: (x: Double, y: Double)? {
        guard let sx, let sy else { return nil }
        return (sx, sy)
    }

    /// Feed one sample. Returns true iff a score event fired.
    @discardableResult
    public func update(_ s: Sample) -> Bool {
        let t = s.t
        justMissed = false // reset every accepted/rejected frame (before any early return below)

        // Order guard: drop out-of-order/duplicate frames (also rejects NaN t).
        guard t > lastT else { return false }
        lastT = t

        // Long-gap discontinuity: reset the track so no crossing is synthesized
        // across a long detector silence. Count/history preserved.
        if lastValidT != -.infinity, t - lastValidT > rule.maxGap {
            resetTrack()
        }

        // Validity gate: finite x/y and (if provided) finite confidence ≥ floor.
        let confOK = s.confidence == nil || (s.confidence!.isFinite && s.confidence! >= rule.minConfidence)
        guard let x = s.x, let y = s.y, x.isFinite, y.isFinite, confOK else {
            return false // a miss — state held, time advanced
        }
        lastValidT = t

        let a = rule.smoothingAlpha
        sx = (sx == nil) ? x : a * x + (1 - a) * sx!
        sy = (sy == nil) ? y : a * y + (1 - a) * sy!

        var fired = false, missed = false
        switch rule.type {
        case .zoneCrossDown:  fired = detectMake(t)
        case .zoneStreak:
            let res = zoneStreak(t)
            fired = (res == .make); missed = (res == .miss)
        case .bounceReversal: fired = bounceReversal(t)
        }

        py = sy
        if fired {
            count += 1
            lastCountT = t
            let ev = ScoreEvent(t: t, count: count,
                                quality: lastQuality?.quality,
                                centerError: lastQuality?.centerError)
            lastQuality = nil
            events.append(ev)
            lastEvent = ev
        } else if missed {
            count = 0 // streak broken — reset, but do NOT set a cooldown (next shot is fair)
            let ev = ScoreEvent(t: t, count: 0, type: "miss")
            events.append(ev)
            lastEvent = ev
            justMissed = true
        }
        return fired
    }

    @discardableResult
    public func feed(_ samples: [Sample]) -> Int {
        for s in samples { _ = update(s) }
        return count
    }

    // Detects a single make through the zone (shared by zoneCrossDown and zoneStreak).
    // Returns true on a make and stashes its swish/rim quality. Behavior is unchanged
    // from the original zoneCrossDown, so the golden parity stays exact.
    private func detectMake(_ t: Double) -> Bool {
        guard let z = rule.zone, let sx = sx, let sy = sy else { return false }
        if z.right <= z.left || z.bottom <= z.top { return false } // degenerate zone never fires

        let inBand = sx >= z.left - rule.xTolerance && sx <= z.right + rule.xTolerance

        // Arm when first seen above the zone and aligned; reset shot-quality
        // tracking on the rising edge so each attempt is classified independently.
        if inBand && sy < z.top {
            if !armed { descentReversals = 0; lastDySign = 0 }
            armed = true; armedT = t
        }
        if armed && t - armedT > rule.armWindow { armed = false }

        // Shot-quality signal: while tracking a descent, count vertical reversals.
        // A clean swish falls monotonically; a rim rattle pops the ball back up
        // before it drops, so a down→up flip is evidence of a rim contact.
        if armed, let py = py {
            let dy = sy - py
            let sign = dy > 1e-4 ? 1 : (dy < -1e-4 ? -1 : 0)
            if sign != 0 {
                if lastDySign == 1 && sign == -1 { descentReversals += 1 }
                lastDySign = sign
            }
        }

        let movingDown = (py == nil) ? true : (sy > py!)
        let crossedBelow = sy > z.bottom
        let cooledDown = t - lastCountT > rule.cooldown

        if armed && inBand && crossedBelow && movingDown && cooledDown && t - armedT <= rule.armWindow {
            armed = false
            lastQuality = classifyShot(z)
            return true
        }
        return false
    }

    private enum StreakResult { case make, miss, none }

    // zoneStreak: detect a make (via detectMake) and, otherwise, a miss — a shot that
    // armed (was aimed at the rim) but then fell more than `missMargin` past the zone
    // without scoring. Returns .make | .miss | .none.
    private func zoneStreak(_ t: Double) -> StreakResult {
        guard let z = rule.zone, let sy = sy else { return .none }
        if z.right <= z.left || z.bottom <= z.top { return .none }
        let wasArmed = armed
        let made = detectMake(t)
        if armed && !wasArmed { attemptActive = true } // a shot just went up at the rim
        if made { attemptActive = false; return .make }
        if attemptActive && sy > z.bottom + rule.missMargin {
            attemptActive = false // the ball fell past the rim without going in
            return .miss
        }
        return .none
    }

    /// Classify a made basket at the crossing point as a clean "swish" or a "rim"
    /// rattle. Pure metadata — it does NOT affect whether/when a score fires, so
    /// the count (and cross-language golden parity) is unchanged. A make is a swish
    /// when it crosses through the central half of the zone with no rim pop-up.
    private func classifyShot(_ z: Zone) -> (quality: String, centerError: Double) {
        let cx = (z.left + z.right) / 2
        let halfW = ((z.right - z.left) / 2 != 0) ? (z.right - z.left) / 2 : 1e-6
        let centerError = abs((sx ?? cx) - cx) / halfW // 0 = dead center, 1 = at the edge
        let clean = centerError <= 0.5 && descentReversals == 0
        return (clean ? "swish" : "rim", centerError)
    }

    private func bounceReversal(_ t: Double) -> Bool {
        guard let sy = sy else { return false }
        let dirSign: Double = (rule.direction == .bottom) ? 1 : -1
        let val = dirSign * sy

        if extremeVal == nil || val > extremeVal! { extremeVal = val }
        if let py = py {
            let towardExtreme = (sy - py) * dirSign
            if towardExtreme > 0 { sawApproach = true }
        }

        let retreat = (extremeVal == nil) ? 0 : extremeVal! - val
        let cooledDown = t - lastCountT > rule.cooldown
        if sawApproach && retreat >= rule.minAmplitude && cooledDown {
            extremeVal = val
            return true
        }
        return false
    }
}

File: web-prototype/js/countingEngine.js
// countingEngine.js
// ----------------------------------------------------------------------------
// Swish Strike — the pure, deterministic counting engine.
//
// This is the brain of the app and the reason the whole thing is testable. It
// has NO knowledge of cameras, the DOM, or any ML model. You feed it a stream of
// ball positions (normalized 0..1 coordinates, y pointing DOWN like image/screen
// space) and it tells you when a "score event" happens — a basketball going
// through the hoop, a soccer ball crossing the goal line, a juggle touch, etc.
//
// The exact same logic exists in Swift at ios/SwishStrikeCore/Sources/SwishStrikeCore/
// CountingEngine.swift. Same input sequence MUST yield the same count in both.
// That contract is guarded three ways: engine.test.mjs, CountingEngineTests.swift,
// and a SHARED golden trace (test/golden.fixtures.json) replayed through both
// languages. Do not change behavior in one without the other.
//
// Robustness (see the guards in update()): out-of-order/duplicate frames are
// dropped, non-finite coordinates are treated as misses, a long detector gap
// resets the track so no phantom crossing is synthesized across an occlusion, and
// a degenerate (zero-area) zone simply never fires.
// ----------------------------------------------------------------------------

export const RuleType = Object.freeze({
  // Ball passes downward through a target zone (hoop, goal, bucket, cornhole).
  ZONE_CROSS_DOWN: 'zoneCrossDown',
  // Like zoneCrossDown, but the count is a CONSECUTIVE streak: a make increments
  // it, a detected miss resets it to 0 (free-throw streak).
  ZONE_STREAK: 'zoneStreak',
  // Ball oscillates and we count amplitude-gated reversals (juggle, dribble, rally).
  BOUNCE_REVERSAL: 'bounceReversal',
});

export const DEFAULTS = Object.freeze({
  smoothingAlpha: 0.5, // EMA factor (0..1). Higher = snappier, lower = smoother.
  minConfidence: 0.30, // detections below this are ignored.
  maxGap: 1.5,         // seconds; a longer gap between valid samples resets the track.
});

const isFiniteNum = (v) => typeof v === 'number' && Number.isFinite(v);

// --- rule builders -----------------------------------------------------------

/**
 * zoneCrossDown rule. `zone` is {left, top, right, bottom} in normalized coords.
 * Fires when the ball, after being seen above the zone and horizontally aligned,
 * crosses below the zone's bottom edge while moving downward — within `armWindow`
 * seconds and respecting `cooldown` seconds between counts.
 */
export function zoneCrossDownRule(zone, opts = {}) {
  return {
    type: RuleType.ZONE_CROSS_DOWN,
    zone: { left: zone.left, top: zone.top, right: zone.right, bottom: zone.bottom },
    xTolerance: opts.xTolerance ?? 0.05,
    armWindow: opts.armWindow ?? 1.5,
    cooldown: opts.cooldown ?? 1.0,
    smoothingAlpha: opts.smoothingAlpha ?? DEFAULTS.smoothingAlpha,
    minConfidence: opts.minConfidence ?? DEFAULTS.minConfidence,
    maxGap: opts.maxGap ?? DEFAULTS.maxGap,
  };
}

/**
 * zoneStreak rule. Detects makes exactly like zoneCrossDown, but the engine count
 * tracks the CONSECUTIVE streak: a make increments it, and a miss (the ball, after
 * being aimed at the target, falls past it — more than `missMargin` below the
 * zone — without scoring) resets it to 0. Used by Free-Throw Streak.
 */
export function zoneStreakRule(zone, opts = {}) {
  const rule = zoneCrossDownRule(zone, opts);
  rule.type = RuleType.ZONE_STREAK;
  rule.missMargin = opts.missMargin ?? 0.18; // how far below the zone counts as a miss
  return rule;
}

/**
 * bounceReversal rule. `direction` is 'bottom' (count troughs — juggle touch /
 * floor bounce) or 'top' (count apexes). `minAmplitude` is the normalized
 * vertical travel required after an extreme before a rep is confirmed (rejects
 * jitter). `cooldown` is the minimum seconds between counts.
 */
export function bounceReversalRule(opts = {}) {
  return {
    type: RuleType.BOUNCE_REVERSAL,
    direction: opts.direction ?? 'bottom',
    minAmplitude: opts.minAmplitude ?? 0.12,
    cooldown: opts.cooldown ?? 0.25,
    smoothingAlpha: opts.smoothingAlpha ?? DEFAULTS.smoothingAlpha,
    minConfidence: opts.minConfidence ?? DEFAULTS.minConfidence,
    maxGap: opts.maxGap ?? DEFAULTS.maxGap,
  };
}

// --- the engine --------------------------------------------------------------

export class CountingEngine {
  constructor(rule) {
    if (!rule || !rule.type) throw new Error('CountingEngine needs a rule');
    this.rule = rule;
    this.reset();
  }

  reset() {
    this.count = 0;
    this.events = [];            // [{t, count, quality?, centerError?}] score events
    this.lastEvent = null;       // most recent score event (incl. shot quality)
    this.justMissed = false;     // true only on the update where a streak miss resolved
    this._lastT = -Infinity;     // last accepted timestamp (order guard)
    this._lastValidT = -Infinity;// last VALID-measurement timestamp (gap guard)
    this._lastCountT = -Infinity;
    this._lastQuality = null;    // shot quality computed at fire, consumed on push
    this._resetTrack();
    return this;
  }

  /** Clears only the per-frame tracking state (not the count/history). */
  _resetTrack() {
    this._sx = null; this._sy = null; this._py = null;
    this._armed = false; this._armedT = -Infinity;     // zoneCrossDown
    this._attemptActive = false;                       // zoneStreak: a shot is in flight
    this._descentReversals = 0; this._lastDySign = 0;  // shot-quality (swish vs rim)
    this._extremeVal = null; this._sawApproach = false; // bounceReversal
  }

  /** Smoothed ball position, or null if nothing seen yet. */
  get position() {
    return this._sx == null ? null : { x: this._sx, y: this._sy };
  }

  /**
   * Feed one sample: { t, x, y, confidence } in normalized coords (y down).
   * Returns true iff a score event fired this sample.
   */
  update(sample) {
    if (!sample) return false;
    const r = this.rule;
    const t = sample.t;
    this.justMissed = false; // reset every accepted/rejected frame (before any early return below)

    // Order guard: drop out-of-order or duplicate frames (also rejects NaN t).
    if (!(t > this._lastT)) return false;
    this._lastT = t;

    // Long-gap discontinuity: a detector silence longer than maxGap means we
    // can't trust the previous point — reset the track so no crossing is
    // synthesized across the gap. (The count/history are preserved.)
    if (this._lastValidT !== -Infinity && t - this._lastValidT > r.maxGap) {
      this._resetTrack();
    }

    // Validity gate: need finite x/y and (if provided) finite confidence ≥ floor.
    const confOK = sample.confidence == null
      || (isFiniteNum(sample.confidence) && sample.confidence >= r.minConfidence);
    if (!(isFiniteNum(sample.x) && isFiniteNum(sample.y) && confOK)) {
      return false; // a miss — state is held, time has advanced
    }
    this._lastValidT = t;

    const a = r.smoothingAlpha;
    this._sx = this._sx == null ? sample.x : a * sample.x + (1 - a) * this._sx;
    this._sy = this._sy == null ? sample.y : a * sample.y + (1 - a) * this._sy;

    let fired = false, missed = false;
    if (r.type === RuleType.ZONE_CROSS_DOWN) fired = this._detectMake(t);
    else if (r.type === RuleType.ZONE_STREAK) {
      const res = this._zoneStreak(t);
      fired = res === 'make'; missed = res === 'miss';
    }
    else if (r.type === RuleType.BOUNCE_REVERSAL) fired = this._bounceReversal(t);

    this._py = this._sy;
    if (fired) {
      this.count += 1;
      this._lastCountT = t;
      const ev = { t, count: this.count };
      if (this._lastQuality) {
        ev.quality = this._lastQuality.quality;
        ev.centerError = this._lastQuality.centerError;
        this._lastQuality = null;
      }
      this.events.push(ev);
      this.lastEvent = ev;
    } else if (missed) {
      this.count = 0; // streak broken — reset, but do NOT set a cooldown (next shot is fair)
      const ev = { t, count: 0, type: 'miss' };
      this.events.push(ev);
      this.lastEvent = ev;
      this.justMissed = true;
    }
    return fired;
  }

  /** Convenience: feed an array, return final count. */
  feed(samples) {
    for (const s of samples) this.update(s);
    return this.count;
  }

  // Detects a single make through the zone (shared by zoneCrossDown and zoneStreak).
  // Returns true on a make and stashes its swish/rim quality. Behavior is unchanged
  // from the original zoneCrossDown, so the golden parity stays exact.
  _detectMake(t) {
    const r = this.rule, z = r.zone;
    if (z.right <= z.left || z.bottom <= z.top) return false; // degenerate zone never fires

    const inBand = this._sx >= z.left - r.xTolerance &&
                   this._sx <= z.right + r.xTolerance;

    // Arm when first seen above the zone and aligned; reset shot-quality tracking
    // on the rising edge so each attempt is classified independently.
    if (inBand && this._sy < z.top) {
      if (!this._armed) { this._descentReversals = 0; this._lastDySign = 0; }
      this._armed = true; this._armedT = t;
    }
    if (this._armed && t - this._armedT > r.armWindow) this._armed = false;

    // Shot-quality signal: while tracking a descent, count vertical reversals.
    // A clean swish falls monotonically; a rim rattle pops the ball back up
    // before it drops, so a down→up flip is evidence of a rim contact.
    if (this._armed && this._py != null) {
      const dy = this._sy - this._py;
      const sign = dy > 1e-4 ? 1 : (dy < -1e-4 ? -1 : 0);
      if (sign !== 0) {
        if (this._lastDySign === 1 && sign === -1) this._descentReversals += 1;
        this._lastDySign = sign;
      }
    }

    const movingDown = this._py == null ? true : this._sy > this._py;
    const crossedBelow = this._sy > z.bottom;
    const cooledDown = t - this._lastCountT > r.cooldown;

    if (this._armed && inBand && crossedBelow && movingDown && cooledDown &&
        t - this._armedT <= r.armWindow) {
      this._armed = false;
      this._lastQuality = this._classifyShot(z);
      return true;
    }
    return false;
  }

  /**
   * Classify a made basket at the crossing point as a clean 'swish' or a 'rim'
   * rattle. Pure metadata — it does NOT affect whether/when a score fires, so the
   * count (and the cross-language golden parity) is unchanged. A make is a swish
   * when it crosses through the central half of the zone with no rim pop-up.
   */
  _classifyShot(z) {
    const cx = (z.left + z.right) / 2;
    const halfW = ((z.right - z.left) / 2) || 1e-6;
    const centerError = Math.abs(this._sx - cx) / halfW; // 0 = dead center, 1 = at the edge
    const clean = centerError <= 0.5 && this._descentReversals === 0;
    return { quality: clean ? 'swish' : 'rim', centerError };
  }

  // zoneStreak: detect a make (via _detectMake) and, otherwise, a miss — a shot
  // that armed (was aimed at the rim) but then fell more than `missMargin` past the
  // zone without scoring. Returns 'make' | 'miss' | null.
  _zoneStreak(t) {
    const z = this.rule.zone;
    if (z.right <= z.left || z.bottom <= z.top) return null;
    const wasArmed = this._armed;
    const made = this._detectMake(t);
    if (this._armed && !wasArmed) this._attemptActive = true; // a shot just went up at the rim
    if (made) { this._attemptActive = false; return 'make'; }
    if (this._attemptActive && this._sy > z.bottom + this.rule.missMargin) {
      this._attemptActive = false; // the ball fell past the rim without going in
      return 'miss';
    }
    return null;
  }

  _bounceReversal(t) {
    const r = this.rule;
    const dirSign = r.direction === 'bottom' ? 1 : -1; // +1 tracks troughs, -1 apexes
    const val = dirSign * this._sy;

    if (this._extremeVal == null || val > this._extremeVal) this._extremeVal = val;
    if (this._py != null) {
      const towardExtreme = (this._sy - this._py) * dirSign;
      if (towardExtreme > 0) this._sawApproach = true;
    }

    const retreat = this._extremeVal == null ? 0 : this._extremeVal - val;
    const cooledDown = t - this._lastCountT > r.cooldown;
    if (this._sawApproach && retreat >= r.minAmplitude && cooledDown) {
      this._extremeVal = val; // re-baseline at the confirmation point
      return true;
    }
    return false;
  }
}

// Single import surface for the app + tests.
export default { RuleType, DEFAULTS, CountingEngine, zoneCrossDownRule, zoneStreakRule, bounceReversalRule };
